fetcher/fetcher.go

```diff
@@ -222,36 +198,44 @@ func (bot *fetcherImpl) Run() {
 	}
 }
 
-// SetContext is part of the Fetcher interface
-func (bot *fetcherImpl) SetContext(ctx context.Context) {
-	bot.ctx = ctx
+// Run is part of the Fetcher interface
+func (bot *fetcherImpl) Run(ctx context.Context) error {
+	// In theory a buffer of size one should be enough, but let's make it bigger.
+	bot.blockQueue = make(chan *rpcs.EncodedBlockCert, 5)
+
+	ctx, cancelFunc := context.WithCancel(ctx)
+	defer cancelFunc()
+
+	ch0 := make(chan error)
+	go func() {
+		ch0 <- bot.processQueue(ctx)
+	}()
+
+	ch1 := make(chan error)
+	go func() {
+		ch1 <- bot.mainLoop(ctx)
+	}()
+
+	select {
+	case err := <-ch0:
+		cancelFunc()
+		<-ch1
+		return fmt.Errorf("Run() err: %w", err)
+	case err := <-ch1:
+		cancelFunc()
+		<-ch0
+		return fmt.Errorf("Run() err: %w", err)
```

Comment:
nit: Disambiguate the errors, more descriptive channel names would also be helpful.
```suggestion
		return fmt.Errorf("Run() processQueue err: %w", err)
	case err := <-ch1:
		cancelFunc()
		<-ch0
		return fmt.Errorf("Run() mainLoop err: %w", err)
```

Reply:
the `err` objects will tell where they come from

Reply:
> the err objects ~will~ **should** tell where they come from

Reply:
But they do. You can check `mainLoop()` and `processQueue()`.